bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,257 @@
+	/// Fee in weth to cover the xcm execution on AH.
+	pub fee: u128,
+	/// The assets
+	pub assets: Vec<Asset>,
+	/// The command originating from the Gateway contract
+	pub xcm: Vec<u8>,
+	/// The claimer in the case that funds get trapped.
+	pub claimer: Option<Vec<u8>>,
+}
+
+/// An asset that will be transacted on AH. The asset will be reserved/withdrawn and placed into
+/// the holding register. The user needs to provide additional xcm to deposit the asset
+/// in a beneficiary account.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum Asset {
+	NativeTokenERC20 {
+		/// The native token ID
+		token_id: H160,
+		/// The monetary value of the asset
+		value: u128,
+	},
+	ForeignTokenERC20 {
+		/// The foreign token ID
+		token_id: H256,
+		/// The monetary value of the asset
+		value: u128,
+	},
+}
+
+/// Reason why a message conversion failed.
+#[derive(Copy, Clone, TypeInfo, PalletError, Encode, Decode, RuntimeDebug)]
+pub enum ConvertMessageError {
+	/// The XCM provided with the message could not be decoded into XCM.
+	InvalidXCM,
+	/// The XCM provided with the message could not be decoded into versioned XCM.
+	InvalidVersionedXCM,
+	/// Invalid claimer MultiAddress provided in payload.
+	InvalidClaimer,
+	/// Invalid foreign ERC20 token ID
+	InvalidAsset,
+	/// The origin could not be added to the interior location of the origin location.
+	InvalidOrigin,
+}
+
+pub trait ConvertMessage {
+	fn convert(message: Message, origin_account: Location) -> Result<Xcm<()>, ConvertMessageError>;
+}
+
+pub struct MessageToXcm<
+	EthereumNetwork,
+	InboundQueuePalletInstance,
+	ConvertAssetId,
+	WethAddress,
+	GatewayProxyAddress,
+> where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	WethAddress: Get<H160>,
+	GatewayProxyAddress: Get<H160>,
+{
+	_phantom: PhantomData<(
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		WethAddress,
+		GatewayProxyAddress,
+	)>,
+}
+
+impl<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		WethAddress,
+		GatewayProxyAddress,
+	> ConvertMessage
+	for MessageToXcm<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		WethAddress,
+		GatewayProxyAddress,
+	>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	WethAddress: Get<H160>,
+	GatewayProxyAddress: Get<H160>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<Xcm<()>, ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
```

Comment:
A possible alternative to failing here could be let the message go through with a empty xcm payload, which will essentially trap the funds on assethub, allowing the user to claim them. Lets discuss this.